### PR TITLE
Disable visibility checks for disclosed contracts and disclosed contract keys

### LINKED ISSUE #14168: Drop warning for divulged contracts for contracts that have been disclosed on submission.
We currently get the warning `Tried to fetch or exercise ${contract.templateId} on contract ${cid.coid} but none of the reading parties [${readers}] are contract stakeholders [${stakeholders}` for contracts that have been disclosed on submission.` This should be dropped.

### LINKED ISSUE #14199: Drop visibility restrictions for locally created contract keys & disclosed contract keys

### PR DESCRIPTION
**Pull request by carlpulley-da**
* Fixes #14168
* Provides a partial solution to #14199 (local contract key visibility changes will be handled on https://github.com/digital-asset/daml/pull/15131)

- [x] Add in predicates to define ~~local contracts & keys and~~ disclosed contracts & keys
- [x] Modify `checkContractVisibility` and `checkKeyVisibility` to side step visibility checks for disclosed contracts & keys
- [x] Add in unit testing for work
- [x] Ensure disabling `checkContractVisibility` and `checkKeyVisibility` changes (e.g. by commenting them out) causes test cases to fail
- [x] Ensure all tests pass

https://github.com/digital-asset/daml/pull/15131 will manage local contract key visibility changes (and so close #14199).

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala [hunk 1]
```diff
@@ -5,26 +5,30 @@ package com.daml.lf
...
+import org.scalatest.Inside
 
-class SpeedyTest extends AnyWordSpec with Matchers {
+import scala.util.{Failure, Success, Try}
 
-  import SpeedyTest._
-  import defaultParserParameters.{defaultPackageId => pkgId}
-
-  def qualify(name: String) = Identifier(pkgId, QualifiedName.assertFromString(name))
+class SpeedyTest extends AnyWordSpec with Matchers with Inside {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala:29]: Boy Scout work: placed val's etc. in a test singleton object

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala [hunk 2]
```diff
@@ -499,21 +456,137 @@ class SpeedyTest extends AnyWordSpec with Matchers {
         )
     }
   }
+
+  "checkContractVisibility" should {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala:460]: Key testing code for Speedy contract & contract key visibility changes

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala [hunk 3]
```diff
@@ -522,14 +595,60 @@ object SpeedyTest {
...
+  implicit def resultEq: Equality[Either[SError, SValue]] = {
     case (Right(v1: SValue), Right(v2: SValue)) => svalue.Equality.areEqual(v1, v2)
     case (Left(e1), Left(e2)) => e1 == e2
     case _ => false
   }
 
-  private def intList(xs: Long*): String =
-    if (xs.isEmpty) "(Nil @Int64)"
-    else xs.mkString(s"(Cons @Int64 [", ", ", s"] (Nil @Int64))")
+  trait VisibilityChecking {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala:609]: Used to define a fixture for contract & contract key visibility testing

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala [hunk 4]
```diff
@@ -136,4 +138,70 @@ private[speedy] object SpeedyTestLib {
       parserParameter: ParserParameters[X]
   ): PureCompiledPackages =
     typeAndCompile(Map(parserParameter.defaultPackageId -> pkg))
+
+  private[lf] object Implicits {
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala:142]: Methods to setup testing data on Speedy machine

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala [hunk 5]
```diff
@@ -136,4 +138,70 @@ private[speedy] object SpeedyTestLib {
...
+        }
+
+        machine
+      }
+
+      private[speedy] def withLocalContractKey(contractId: ContractId, key: GlobalKey): Machine = {
+        machine.ledgerMode match {
+          case ledger: OnLedger =>
+            ledger.ptx = ledger.ptx.copy(
+              contractState = ledger.ptx.contractState.copy(
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTestLib.scala:178]: Not convinced this is the best way to modify the contract state machine, but it at least allows the scenarios we're interested in testing to be mocked up

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 6]
```diff
@@ -2438,9 +2456,17 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
           )
           inside(res) {
             case Success(
-                  Left(SErrorDamlException(IE.ContractKeyNotVisible(_, key, _, _, _)))
+                  Left(
+                    SErrorDamlException(
+                      IE.FailedAuthorization(
+                        _,
+                        FetchMissingAuthorization(`T`, _, stakeholders, authParties),
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:2463]: As local contracts now always pass visibility checking, we now fail due to authorisation issues

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 7]
```diff
@@ -2984,9 +3010,17 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
           )
           inside(res) {
             case Success(
-                  Left(SErrorDamlException(IE.ContractKeyNotVisible(_, key, _, _, _)))
+                  Left(
+                    SErrorDamlException(
+                      IE.FailedAuthorization(
+                        _,
+                        LookupByKeyMissingAuthorization(`T`, _, maintainers, authParties),
```
Comment (carlpulley-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:3017]: As local contracts now always pass visibility checking, we now fail due to authorisation issues

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala [hunk 8]
```diff
@@ -384,20 +384,16 @@ final class ContractKeysIT extends LedgerTestSuite {
 
   test(
     "CKLocalKeyVisibility",
-    "Visibility should be checked for fetch-by-key/lookup-by-key of contracts created in the current transaction",
+    "Visibility should *not* be checked for fetch-by-key/lookup-by-key of contracts created in the current transaction",
```
Comment (carlpulley-da) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/ContractKeysIT.scala:387]: local contract keys are no longer subjected to visibility checking

File: compiler/damlc/tests/daml-test-files/ContractKeyNotVisible.daml [hunk 9]
```diff
@@ -118,11 +116,4 @@ localFetch = scenario do
   cid <- submit p1 $ create (LocalKeyVisibility p1 p2)
   submit p2 $ exercise cid LocalFetch
 
-localLookupFetchMustFail = scenario do
```
Comment (carlpulley-da) [inline compiler/damlc/tests/daml-test-files/ContractKeyNotVisible.daml:121]: Removed as:
- lookup and fetch are tested above
- lookup followed by fetch will fail due to contract being consumed

File: compiler/damlc/tests/src/DA/Test/ScriptService.hs [hunk 10]
```diff
@@ -919,25 +919,17 @@ main =
                     , "  p2 <- allocateParty \"p2\""
                     , "  cid <- submit p1 $ createCmd (LocalKeyVisibility p1 p2)"
                     , "  submit p2 $ exerciseCmd cid LocalFetch"
-                    , "localLookupFetchMustFail = do"
```
Comment (carlpulley-da) [inline compiler/damlc/tests/src/DA/Test/ScriptService.hs:922]: lookup and fetch testing is done above. A lookup and fetch test now always fails on the fetch as contract has been consumed.

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ContractKeysIT.scala [hunk 11]
```diff
@@ -0,0 +1,554 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (carlpulley-da) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ContractKeysIT.scala:1]: Mostly a copy of the v1_8/ContractKeysIT.scala file (key changes marked below)

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ContractKeysIT.scala [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+      assertGrpcError(
+        failureAfterConsuming,
+        LedgerApiErrors.CommandExecution.Interpreter.LookupErrors.ContractKeyNotFound,
+        Some("dependency error: couldn't find key"),
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
```
Comment (carlpulley-da) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ContractKeysIT.scala:385]: This is the only test in this file that differs from v1_8/ContractKeysTest.scala. As elsewhere, local contract keys no longer under go visibility checking, hence this test has now simplified.

File: compiler/damlc/tests/daml-test-files/ContractKeyNotVisible.daml [hunk 13]
```diff
@@ -1,6 +1,8 @@
--- @ERROR range=16:1-16:10; Attempt to fetch, lookup or exercise a key associated with a contract
--- @ERROR range=59:1-59:15; Attempt to fetch, lookup or exercise a key associated with a contract
--- @ERROR range=77:1-77:12; Attempt to fetch, lookup or exercise a key associated with a contract
+-- @ERROR range=18:1-18:10; Attempt to fetch, lookup or exercise a key associated with a contract
+-- @ERROR range=61:1-61:15; Attempt to fetch, lookup or exercise a key associated with a contract
+-- @ERROR range=79:1-79:12; Attempt to fetch, lookup or exercise a key associated with a contract
+-- @ERROR range=109:1-109:12; Attempt to fetch, lookup or exercise a key associated with a contract
+-- @ERROR range=115:1-115:11;  Attempt to fetch, lookup or exercise a key associated with a contract
```
Comment (remyhaemmerle-da) [inline compiler/damlc/tests/daml-test-files/ContractKeyNotVisible.daml:5]:
```suggestion
-- @ERROR range=18:1-18:10; Attempt to fetch, lookup or exercise a key associated with a contract not visible to the committer
-- @ERROR range=61:1-61:15; Attempt to fetch, lookup or exercise a key associated with a contract not visible to the committer
-- @ERROR range=79:1-79:12; Attempt to fetch, lookup or exercise a key associated with a contract not visible to the committer
-- @ERROR range=109:1-109:12; Attempt to fetch, lookup or exercise a key associated with a contract not visible to the committer
-- @ERROR range=115:1-115:11;  Attempt to fetch, lookup or exercise a key associated with a contract not visible to the committer


```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 14]
```diff
@@ -258,6 +269,10 @@ private[lf] object Speedy {
     private[speedy] def contractIdByKey(keyHash: crypto.Hash): Option[SValue.SContractId] =
       keyMap.get(keyHash)
 
+    private[speedy] def isValidEntry(key: GlobalKey, contractId: V.ContractId): Boolean = {
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:272]: what do you mean by entry ?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 15]
```diff
@@ -863,18 +878,26 @@ private[lf] object Speedy {
...
+            val readers = (actAs union readAs).mkString(",")
+            val stakeholders = contract.stakeholders.mkString(",")
+            this.warningLog.add(
+              Warning(
+                commitLocation = onLedger.commitLocation,
+                message =
+                  s"Tried to fetch or exercise ${contract.templateId} on contract ${cid.coid} "
+                    + s"but none of the reading parties [$readers] are contract stakeholders [$stakeholders]. "
+                    + "Use of divulged contracts is deprecated and incompatible with pruning. "
+                    + s"To remedy, add one of the readers [$readers] as an observer to the contract.",
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:897]:
```suggestion
                     s"""Tried to fetch or exercise ${contract.templateId} on contract ${cid.coid}
                     | but none of the reading parties [$readers] are contract stakeholders [$stakeholders]. 
                     | Use of divulged contracts is deprecated and incompatible with pruning. 
                     | To remedy, add one of the readers [$readers] as an observer to the contract.""".stripMargin,
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 16]
```diff
@@ -884,26 +907,33 @@ private[lf] object Speedy {
...
+    ): Control = {
+      // For disclosed contract keys, we do not perform visibility checking
+      if (onLedger.isDisclosedContractKey(coid, gkey)) {
+        handleKeyFound(this, coid)
+      } else {
+        onLedger.getCachedContract(coid) match {
+          case Some(cachedContract) =>
+            val stakeholders = cachedContract.signatories union cachedContract.observers
+            onLedger.visibleToStakeholders(stakeholders) match {
+              case SVisibleToStakeholders.Visible =>
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:919]: :+1:

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala [hunk 17]
```diff
@@ -455,7 +456,7 @@ class EvaluationOrderTest extends AnyFreeSpec with Matchers with Inside {
   // is rejected in both cases “only” change the error message which is relatively harmless.
   // Specifically this means that we need to test ordering of catchable errors
   // relative to other catchable errors and other non-catchable errors but we don’t
-  // need to check ordering of non-catchable errors relative to other non-cachable errors.
+  // need to check ordering of non-catchable errors relative to other non-catchable errors.
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/EvaluationOrderTest.scala:459]: :bow:

File: daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala [hunk 18]
```diff
@@ -499,21 +456,148 @@ class SpeedyTest extends AnyWordSpec with Matchers {
...
+                interpretation.Error.ContractKeyNotVisible(
+                  `localContractId`,
+                  `localContractKey`,
+                  _,
+                  _,
+                  _,
+                )
+              )
+            ) =>
+          succeed
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/SpeedyTest.scala:503]: I did not know this. Good to known.

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 19]
```diff
@@ -163,6 +163,17 @@ private[lf] object Speedy {
     private[lf] def incompleteTransaction: IncompleteTransaction = ptx.finishIncomplete
     private[lf] def nodesToString: String = ptx.nodesToString
 
+    private[speedy] def isDisclosedContract(contractId: V.ContractId): Boolean = {
+      ptx.disclosedContracts.find(_.contractId.value == contractId).nonEmpty
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:167]: I'll add a data structure to PartialTransaction that allows this to be tested in constant time. If `pix` wasn't mutable, we could have it here. :-(

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 14 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:272]: Renamed to the less generic `isValidDisclosedContractKeyEntry`

**Pull request merged**